### ISSUE TITLE
Compiling native extension fails when having `psych >= 5`

### ISSUE THREAD

**Issue by manuelvanrijn** [appsignal/appsignal-ruby#904]
Timebox: 0.5 days

## Describe the bug

After we've downgraded back to psych 4 compilation worked again.

## To Reproduce

We are on `ruby 3.1.3p185 (2022-11-24 revision 1a6b16756e) [aarch64-linux-musl]`

```
/app # gem install psych:5
/app # gem install appsignal:3.2.2
```

Error output:
```
Building native extensions. This could take a while...
ERROR:  Error installing appsignal:
	ERROR: Failed to build gem native extension.

    current directory: /usr/local/bundle/gems/appsignal-3.2.2/ext
/usr/local/bin/ruby -I /usr/local/lib/ruby/3.1.0 extconf.rb
/usr/local/lib/ruby/3.1.0/psych.rb:455:in `parse_stream': undefined method `parse' for #<Psych::Parser:0x0000ffffa35c9628 @handler=#<Psych::Handlers::DocumentStream:0x0000ffffa35c9b78 @stack=[], @last=nil, @root=nil, @start_line=nil, @start_column=nil, @end_line=nil, @end_column=nil, @block=#<Proc:0x0000ffffa35c96f0 /usr/local/lib/ruby/3.1.0/psych.rb:399>>, @external_encoding=0> (NoMethodError)

      parser.parse yaml, filename
            ^^^^^^
	from /usr/local/lib/ruby/3.1.0/psych.rb:399:in `parse'
	from /usr/local/lib/ruby/3.1.0/psych.rb:323:in `safe_load'
	from /usr/local/lib/ruby/3.1.0/psych.rb:369:in `load'
	from /usr/local/bundle/gems/appsignal-3.2.2/ext/base.rb:12:in `<top (required)>'
	from <internal:/usr/local/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from <internal:/usr/local/lib/ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:85:in `require'
	from extconf.rb:1:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /usr/local/bundle/gems/appsignal-3.2.2 for inspection.
Results logged to /usr/local/bundle/extensions/aarch64-linux-musl/3.1.0/appsignal-3.2.2/gem_make.out
```

**Comment by jgreidy** [comment]
Thank you. This solved the problem for us. Ours was on Jenkins, using rvm, where psych 5.x somehow got installed, and was then secretly used to build appsignal. 
gem uninstall psych  
to list the versions, then
gem uninstall psych -v 5.0.1
to get rid of them
then 
gem install psych -v 4.0.6

**Comment by tombruijn** [comment]
Thanks for reporting this issue, but I am unable to reproduce with the commands you provided.

The error is triggered when the `Psych::Parser#parse` method is not found. In Psych 4, this method is [defined by the Psych C-extension](https://github.com/ruby/psych/blob/2c3708e0a483c6d44ebddaff0b524166f3e7bc78/ext/psych/psych_parser.c#L561). This error could occur when the C-extension failed to install, but that version of the psych gem is installed when Ruby is installed. I would expect the entire Ruby version to fail to install then.

In Psych 5 the `parse` method is [defined in the Ruby gem](https://github.com/ruby/psych/blob/1f23e6e7f0ab4a6efab598c1ee528bb52d40ee51/lib/psych/parser.rb#L61-L63). This method calls a [C function called `_native_parse`](https://github.com/ruby/psych/blob/1f23e6e7f0ab4a6efab598c1ee528bb52d40ee51/ext/psych/psych_parser.c#L547), which is the renamed version of the `parse` function.

My best guess is that the wrong C-extension is loaded when the psych 5 gem is installed in your case. That particular scenario I have not been able to reproduce. I'm using chruby, not rvm, but my colleague using rvm can't reproduce this issue either.

To only way I get this error, is by installing psych using the `gem install psych --default` command (which installs version 5.0.1). That will break the entire install though so don't run it unless you like reinstalling Ruby. I couldn't even run `gem list` and `gem uninstall` after that.
My guess is that the psych 5 gem C-extension then becomes the default psych C-extension to load, but the `gem install` command (and others) still use the psych 4 codebase, shipped with Ruby itself.

Was psych 5 installed using `gem install psych` or some other way originally?

The issue itself isn't in the AppSignal gem, so I can't fix it.

**Comment by tijn** [comment]
@tombruijn, for us, psych 5 gets installed because rdoc depends on it. Bundler simply tries to install the latest version.

Appsignal doesn't list psych in the gemspec (because it's already installed with Ruby?) but rdoc explicitly specifies their dependency like such:

```ruby
s.add_dependency 'psych', '>= 4.0.0'
```

I think that this issue can be resolved if AppSignal too could write out the dependency on psych but with an extra requirement since AppSignal appears to be incompatible with the newest version:

```ruby
s.add_dependency 'psych', '>= 4.0.0', '<= 5.0.0'
```

This will also stop dependabot from trying to upgrade libraries that cannot be upgraded.

**Comment by manuelvanrijn** [comment]
@tombruijn Sorry for the daily. Here's an example how you can reproduce the problem:

1. `docker run --rm -ti ruby:3.1-alpine3.15 apk add --update --no-cache build-base && sh`
2. `vi Gemfile`
```ruby
source 'https://rubygems.org'

gem 'sdoc'
gem 'appsignal', '3.2.2'
```
3. `bundle install`
4. `vi Gemfile` and change the version number of appsignal to 3.3.1. So:
```ruby
source 'https://rubygems.org'

gem 'sdoc'
gem 'appsignal', '3.3.1'
```
5. Run `bundle install`

**Comment by LeipeLeon** [comment]
crap... @manuelvanrijn, you beat me to it! :P

Created an repo which also can reproduces the error: https://github.com/Memoriam-tv/appsignal-ruby-test-report/actions

**Comment by mvz** [comment]
For me, just having psych 5 installed breaks the appsignal installation, even when psych 5 is not part of the bundle.

**Comment by vdbilt** [comment]
Check, update to 3.3.1 is failing in all our repo's because of this.

**Comment by tombruijn** [comment]
Thanks for the extra information!
I find it amazing I still can't reproduce this locally. Even with the instructions for the alpine container. That installs fine for me 🤷‍♂️

I'll try some more with the test repo.

We may just need to add psych as a dependency for a while then. It's on my wishlist to remove the YAML dependency. It has broken so many times before already. ([Private issue for removing YAML](https://github.com/appsignal/appsignal-agent/issues/736).)

**Comment by tombruijn** [comment]
I can add psych as a dependency, but this causes the following:

- If I lock it to psych version 4, you can't use it in apps using psych 5.
- If I lock it to psych version 5, you can't use it in apps using psych 4.

Either would not be ideal.

If I remove our YAML reading from the gem installation entirely, surprise! It reads YAML when we [try to figure out Ruby's proxy settings](https://github.com/appsignal/appsignal-ruby/blob/7462000220c799ae44e07247bc86a5642521cca8/ext/base.rb#L190), using `Gem.configuration[:http_proxy]` and fails on that instead.

I think this is a bug in Ruby. It looks to me it's [loading the wrong psych extension and Ruby code combination](https://github.com/appsignal/appsignal-ruby/issues/904#issuecomment-1369041729).

I'll try to double sure confirm this theory and write up a report and/or try to see if we can figure out the proxy setting another way and go the YAML-less route.

**Comment by mvz** [comment]
Adding the dependency won't be enough, since having psych 5 installed without it being part of the bundle is enough to trigger the problem.

**Comment by tombruijn** [comment]
You're right. It doesn't help to specify psych 4 in the bundle.

This is not a fun bug. I think I'm going for the non-YAML route and remove the http proxy thing that deep down also parses some YAML. That way it will work in all scenarios, expect when you rely on the Ruby gems http proxy config, for which people can then specify the HTTP_PROXY env var as a workaround.

**Comment by tombruijn** [comment]
Partial fix sent in as a PR: https://github.com/appsignal/appsignal-ruby/pull/913
That still has the issue with the `Gem.configuration[:http_proxy]` line causing a similar psych issue later on. I'll address that in another PR.

**Comment by tijn** [comment]
@tombruijn Oh, right! My "solution" only works if you have an isolated app with its own gems (like we have in our Docker setup). I completely forgot about other situations :-)

**Comment by tombruijn** [comment]
@tijn no worries. I also forgot to mention I tested our YAML-less install on Ruby 3.2 with pysch 5 and that does work. Meaning upgrading Ruby to version 3.2 is also an option.

We are compatible with psych 5, it's just that any Ruby version with version 4 in the standard library doesn't work, which is most Rubies.

I've submitted another part of the fix in PR #914. The two PRs #913 and #914, when merged, should fix it.

**Comment by tombruijn** [comment]
I've released the fix for this in AppSignal for Ruby gem 3.3.2.

Only caveat: We try to read the `.gemrc` file, but if this can't be done because of this issue it's ignored. Instead use the `HTTP_PROXY` env var.

Please try out this new version and let us know if it works or not!

**Comment by tijn** [comment]
I just installed the newest gem in all our apps @tombruijn and there's no trace of this issue anymore.

**Comment by LeipeLeon** [comment]
Looks good on my deploys! cheers!

**Comment by tombruijn** [comment]
I've reported it with Ruby upstream: https://bugs.ruby-lang.org/issues/19371

**Comment by tombruijn** [comment]
This issue was fixed in RubyGems 3.4: https://github.com/rubygems/rubygems/pull/6490
Please update to the latest RubyGems with: `gem update --system` if you run into this issue.
